### PR TITLE
Don't replace FloorDiv with floor in simplify, do simplifications for divisible exprs

### PR DESCRIPTION
**Pull request by ngimel**
I don't see why `floor` is better than `FloorDiv` and solve with `FloorDiv` doesn't work anyway (the solution wouldn't be unique even if it worked).

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -1422,6 +1420,7 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
         Evaluates the result of an eq call. If true, uses information to
         simplify shapes (i.e. a == b or a % 5 == 0)
         """
+        print("EXPR", expr)
```
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:1423]: reminder to remove this later

**Comment by nkaretnikov** [comment]
I've added some CI tags to see how this affects models. Will take a look closely once everything is green (or broken).

**Review by jansel — APPROVED** [review]

**Review by malfet — APPROVED** [review]

**Review by Chillee — CHANGES_REQUESTED** [review]
I'm somewhat confused here - I don't see why this is any better. FloorDiv is still not simplifiable by us. Perhaps we could just add a pass in simplify that simplifies FloorDiv patterns? 

I also have another (hacky) idea - what if we replaced `FloorDiv(a, b)` with `a/b` for the duration of the simplify, and then remap `a/b` back to `FloorDiv(a, b)` if it doesn't simplify?

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1364,8 +1364,6 @@ def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
             div_replacements = {}
             for atom in expr.atoms(FloorDiv):
                 base, divisor = atom.args
-                if self.replace(base % divisor) in self.divisible:
```
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/fx/experimental/symbolic_shapes.py:1367]: We are not doing anything with divisibility anyway, floor(a/b) is correct replacement even for non-divisible a and b. Replacing for the duration of simplify and trying to cancel division with multiplication is good, I'll add this.

[the remark this answers is not in the document]
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:1367]: Sympy implicitly simplifies a bunch of divisibility related things, which is why this is here.

[the remark this answers is not in the document]
Comment (ngimel) [reply torch/fx/experimental/symbolic_shapes.py:1367]: Does sympy do anything for `a/b` that FloorDiv doesn't do in eval (like dividing by gcd)?

**Comment by jansel** [comment]
> I also have another (hacky) idea - what if we replaced `FloorDiv(a, b)` with `a/b` for the duration of the simplify, and then remap `a/b` back to `FloorDiv(a, b)` if it doesn't simplify?

This is a bad idea.  Don't do it.

`a//b*b` will become `a`, which is incorrect if there is a remainder in the divide.

**Comment by ngimel** [comment]
> `a//b*b` will become `a`, which is incorrect if there is a remainder in the divide.

We should do it only if a is divisible by b (we have divisibility info in shape env).

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1559]: For reference: `floor` has some things that are different compared to `FloorDiv`. See `class floor` [here](https://github.com/sympy/sympy/blob/fb7c59c1e5f15b625bc3b777c55e26872ec787b9/sympy/functions/elementary/integers.py#L95):
-  more `_eval` methods
-  an eval rule that is triggered during equality (`_eval_is_eq`).

It's hard to say whether these matter much _for us_ in practice, though. However, waiting on cases that are hard to `solve` is a real problem that we have now, so I guess it makes sense to do this change just to fix that.

As for divisibility, we could try to do optimizations somewhere else, e.g., in custom `Mul`, but that's separate. Natalia correctly pointed out that we're not doing much with divisibility here anyway.

We can always go back or do something else if there are problems, so I see no reason to block this.

**Review by nkaretnikov — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1559]: FYI: floor/mul optimization PR is here: https://github.com/pytorch/pytorch/pull/95120 (WIP)

[the remark this answers is not in the document]
Comment (ngimel) [reply torch/fx/experimental/symbolic_shapes.py:1559]: I've added floor/mul optimization to this PR, it works for all tricky cases in swin.

**Pull request closed without merging**